File: programmers/HEAP/더 맵게.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = -1;
    priority_queue<int, vector<int>, greater<int> > leo_dishes;
    for(int dish : scoville){
        leo_dishes.push(dish);
    }
    
    int now, next, cnt;
    for(cnt = 0; leo_dishes.size() > 1 ; ++cnt){
        now = leo_dishes.top();
        leo_dishes.pop();
        next = leo_dishes.top();
        leo_dishes.pop();
        if(now >= K)    break;
        leo_dishes.push(now + next * 2);        
    }
    if(leo_dishes.top() >= K)   return cnt;
    else                        return -1;
}

File: programmers/HEAP/더_맵게.java
package com.kt.aicare.common.util.programmers.HEAP;
import java.util.*;

public class 더_맵게 {
  void offerAll(int[] elems, PriorityQueue<Integer> target) {
    for (int elem : elems) {
      target.offer(elem);
    }
  }

  void mixFoods(PriorityQueue<Integer> container) {
    if (container.size() < 2) {
      return;
    }
    int food1 = container.poll();
    int food2 = container.poll();

    container.offer(food1 + food2 * 2);
  }

  public int solution(int[] scoville, int K) {
    PriorityQueue<Integer> pq = new PriorityQueue<>();
    offerAll(scoville, pq);

    for(int cnt = 0; cnt < scoville.length; cnt++) {
      if (pq.peek() >= K) {
        return cnt;
      }
      mixFoods(pq);
    }

    return -1;
  }

}
